### PR TITLE
Build with docker compose

### PR DESCRIPTION
**Pull request by chelnak**
### Context
Switches the az pipelines build process to use docker-compose.

We now tag each image explicitly. This causes the cache_from command to fail as no image available with the latest tag on docker hub.

### Changes proposed in this pull request
* Add -T switch to exec commands fix TTY error
* Align tagging with Travis steps
* Use --out parameter to push tests to an xml file on the host
* Remove docker-compose pull from build steps to avoid dependencies on latest
* Comment out cache_from

### Guidance to review
* Review changes / Make suggestions on how the process should work

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -10,9 +10,9 @@ services:
   web:
     build:
       context: .
-      cache_from:
-        - batdevcontainerregistry.azurecr.io/manage-courses-backend:latest
-    image: batdevcontainerregistry.azurecr.io/manage-courses-backend:latest
+      #cache_from:
+      #  - govukdfe/manage-courses-backend-poc:latest
```
Comment (timabell) [inline docker-compose.yml]: just delete it, we can always put it back

File: azure-pipelines.yml [hunk 2]
```diff
@@ -15,26 +15,30 @@ variables:
...
+    docker-compose exec -T web /bin/sh -c 'bundle exec rake spec SPEC_OPTS="--format RspecJunitFormatter --out ./rspec-results.xml"'
   displayName: 'Execute tests'
 
 - script: |
-    docker run "$(tag)" /bin/sh -c "bundle exec govuk-lint-ruby app config db lib spec --format clang"
+    docker-compose exec -T web /bin/sh -c "bundle exec govuk-lint-ruby app config db lib spec --format clang"
   displayName: 'Execute linters'
 
 - script: |
     echo $password | docker login --username $(dockerHubUsername) --password-stdin
```
Comment (defong) [inline azure-pipelines.yml]: Echoing password? Are you sure?

File: azure-pipelines.yml [hunk 3]
```diff
@@ -15,26 +15,30 @@ variables:
...
-    echo "##vso[task.setvariable variable=tag;]$tag"
+    echo "##vso[task.setvariable variable=docker_path;]$docker_path"
   displayName: 'Set version number'
 
 - script: |
-    docker build -f Dockerfile -t $(tag) .
-  displayName: 'Build image'
+    docker-compose build
+    docker-compose up --no-build -d
+    docker-compose exec -T web /bin/sh -c "./wait-for-command.sh -c 'nc -z db 5432' -s 0 -t 20"
```
Comment (defong) [inline azure-pipelines.yml]: sh vs ash for shell?

File: docker-compose.yml (hunk 1 above)
Comment (chelnak) [reply docker-compose.yml]: Removed

File: azure-pipelines.yml (hunk 2 above)
Comment (chelnak) [reply azure-pipelines.yml]: https://docs.docker.com/engine/reference/commandline/login/

See **Provide a password using STDIN**

**Comment by chelnak** [comment]
@defong Interesting point re what shell to use.

ash is used inside docker-compose.yml however sh is used in the build (travis and az pipelines).

Probably a wider conversation?

**Comment by defong** [comment]
> @defong Interesting point re what shell to use.
> 
> ash is used inside docker-compose.yml however sh is used in the build (travis and az pipelines).
> 
> Probably a wider conversation?

Let's use ps...

Can we just sh across the board, I don't think ash give us anything particaular

File: azure-pipelines.yml [hunk 4]
```diff
@@ -11,30 +11,41 @@ pool:
 
 variables:
   imageName: 'manage-courses-backend-poc'
+  dockerOverride: 'docker-compose -f docker-compose.yml -f docker-compose.az.yml'
```
Comment (timabell) [inline azure-pipelines.yml]: I'm not sure I understand what this does, which might be fine till I need to look after it :-)
Perhaps a comment on the line above explaining what it's for and why it's there, no need to repeat things I can read in the docs, though a link might be handy.

Comment (chelnak) [reply azure-pipelines.yml]: Happy to link to the docs. This will go away when travis goes though.

Comment (chelnak) [reply azure-pipelines.yml]: https://github.com/DFE-Digital/manage-courses-backend/pull/133/commits/15104ec9004137b8cee9aee2acbb7fc65b7b4ce4

**Pull request merged**